s3fs/tests/test_s3fs.py

```diff
@@ -2464,6 +2465,34 @@ def test_exists_isdir(s3):
     assert not s3.isdir(bad_path)
 
 
+def test_exists_raises_on_connection_error(monkeypatch):
+    # Ensure that we raise a ConnectionError instead of returning False if we
+    # are not actually able to connect to storage, by setting a fake proxy and
+    # then re-creating the S3FileSystem instance.
+    monkeypatch.setenv('https_proxy', 'https://fakeproxy.127.0.0.1:8080')
+    S3FileSystem.clear_instance_cache()
+    s3 = S3FileSystem(anon=False, client_kwargs={"endpoint_url": endpoint_uri})
+    s3.invalidate_cache()
+    with pytest.raises(EndpointConnectionError):
+        s3.exists('this-bucket-does-not-exist/')
+
+
+def test_exists_bucket_nonexistent_or_no_access(caplog):
+    # Ensure that a warning is raised and False is returned if querying a
+    # bucket that might either not exist or be private.
+    fs = s3fs.S3FileSystem(key="asdfasfsdf", secret="sadfdsfds")
+    assert not fs.exists("s3://this-bucket-might-not-exist/")
+    assert caplog.records[0].levelname == "WARNING"
+    assert "doesn't exist or you don't have access" in caplog.text
+
+
+def test_exists_bucket_nonexistent(s3, caplog):
+    # Ensure that NO warning is raised and False is returned if checking bucket
+    # existance when we have full access.
```

Comment:
Well, the bucket might still exist, but it's not one we can see. We happen to know for the sake of the test that it doesn't exist.

Reply:
Not sure whether this is true for all S3 systems, but at least for our internal systems, I get a "PermissionError" when trying to access a bucket I cannot see (because I don't have permission to see it) but a `FileNotFoundError` like in this test when I would have permission to see, but it simply doesn't exist. This is why in the PR change, the warning would only be raised in case of `PermissionError` but not in case of `FileNotFoundError`. Should I change it to return the warning in both cases?